File: ruby/lib/runapi/kling/resources/image_to_video.rb
# frozen_string_literal: true

module RunApi
  module Kling
    module Resources
      # Kling image-to-video resource.
      # Generate videos from an input image.
      class ImageToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/kling/image_to_video"

        RESPONSE_CLASS = Types::ImageToVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedImageToVideoResponse
        V26_MODEL = "kling-v2.6"
        V3_TURBO_MODEL = "kling-v3-turbo-image-to-video"
        V3_TURBO_UNSUPPORTED_FIELDS = %i[
          aspect_ratio
          negative_prompt
          cfg_scale
          last_frame_image_url
        ].freeze

        def initialize(http)
          @http = http
        end

        # Generate an image-to-video task and wait until complete.
        #
        # @param params [Hash] image-to-video parameters
        # @return [RunApi::Kling::Types::CompletedImageToVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an image-to-video task.
        #
        # @param params [Hash] image-to-video parameters
        # @return [RunApi::Kling::Types::ImageToVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get image-to-video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Kling::Types::ImageToVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          reject_unsupported_v3_turbo_fields!(params)
          validate_contract!(CONTRACT["image-to-video"], params)

          # Bespoke: last_frame_image_url is only allowed for select models
          # (model-gating, not expressible as a contract enum/required rule).
          model = param(params, :model)
          last_frame_image_url = param(params, :last_frame_image_url)
          if model == V26_MODEL
            validate_v26_params!(params, last_frame_image_url)
          elsif last_frame_image_url && !%w[kling-v2.5-turbo-image-to-video-pro kling-v2.1-pro].include?(model)
            raise Core::ValidationError, "last_frame_image_url is only supported by kling-v2.5-turbo-image-to-video-pro and kling-v2.1-pro"
          end
        end

        def validate_v26_params!(params, last_frame_image_url)
          mode = param(params, :mode) || "std"
          if param(params, :enable_sound) == true && mode != "pro"
            raise Core::ValidationError, "enable_sound requires mode pro for #{V26_MODEL}"
          end
          return unless last_frame_image_url

          raise Core::ValidationError, "last_frame_image_url requires mode pro for #{V26_MODEL}" unless mode == "pro"

          duration_seconds = param(params, :duration_seconds) || 5
          return if duration_seconds.to_i == 5

          raise Core::ValidationError, "last_frame_image_url requires duration_seconds 5 for #{V26_MODEL}"
        end

        def reject_unsupported_v3_turbo_fields!(params)
          return unless param(params, :model) == V3_TURBO_MODEL

          field = V3_TURBO_UNSUPPORTED_FIELDS.find { |candidate| field_present?(params, candidate) }
          raise Core::ValidationError, "#{field} is not supported by #{V3_TURBO_MODEL}" if field
        end
      end
    end
  end
end

File: python/src/runapi/kling/resources/image_to_video.py
"""Kling image-to-video resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedImageToVideoResponse,
    ImageToVideoResponse,
)
from .o1_reference_validation import O1_MODEL, validate_kling_o1_references

V26_MODEL = "kling-v2.6"
V3_OMNI_MODEL = "kling-v3-omni"
V3_TURBO_MODEL = "kling-v3-turbo-image-to-video"
V3_TURBO_UNSUPPORTED_FIELDS = (
    "aspect_ratio",
    "negative_prompt",
    "cfg_scale",
    "last_frame_image_url",
)


class ImageToVideo(Resource):
    """Generate videos from an input image with Kling models."""

    ENDPOINT = "/api/v1/kling/image_to_video"

    RESPONSE_CLASS = ImageToVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedImageToVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Generate a video from an image and poll until it completes.

        Args:
            **params: Image-to-video parameters (model, prompt, first_frame_image_url, ...).

        Returns:
            The completed task with videos.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an image-to-video task and return immediately with an ``id``.

        Args:
            **params: Image-to-video parameters (model, prompt, first_frame_image_url, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of an image-to-video task.

        Args:
            id: The task id.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._reject_unsupported_v3_turbo_fields(params)
        self._validate_contract(CONTRACT["image-to-video"], params)
        validate_kling_o1_references(params)

        # Bespoke last-frame rules that the generated contract cannot express.
        model = params.get("model")
        last_frame_image_url = params.get("last_frame_image_url")
        if model == V26_MODEL:
            self._validate_v26_params(params, last_frame_image_url)
        elif last_frame_image_url and model == V3_OMNI_MODEL:
            if params.get("duration_seconds", 5) != 5:
                raise ValidationError(
                    f"last_frame_image_url requires duration_seconds 5 for {V3_OMNI_MODEL}"
                )
        elif last_frame_image_url and model not in (
            O1_MODEL,
            "kling-v2.5-turbo-image-to-video-pro",
            "kling-v2.1-pro",
        ):
            raise ValidationError(
                "last_frame_image_url is only supported by "
                "kling-v2.5-turbo-image-to-video-pro and kling-v2.1-pro"
            )

    def _validate_v26_params(
        self, params: Dict[str, Any], last_frame_image_url: Any
    ) -> None:
        mode = params.get("mode") or "std"
        if params.get("enable_sound") is True and mode != "pro":
            raise ValidationError(f"enable_sound requires mode pro for {V26_MODEL}")
        if not last_frame_image_url:
            return
        if mode != "pro":
            raise ValidationError(f"last_frame_image_url requires mode pro for {V26_MODEL}")
        if params.get("duration_seconds", 5) != 5:
            raise ValidationError(
                f"last_frame_image_url requires duration_seconds 5 for {V26_MODEL}"
            )

    def _reject_unsupported_v3_turbo_fields(self, params: Dict[str, Any]) -> None:
        if params.get("model") != V3_TURBO_MODEL:
            return

        for field in V3_TURBO_UNSUPPORTED_FIELDS:
            if self._field_present(params, field):
                raise ValidationError(f"{field} is not supported by {V3_TURBO_MODEL}")
